torch/distributed/elastic/rendezvous/c10d_rendezvous_backend.py

```diff
@@ -0,0 +1,195 @@
+
+        base64_state: bytes = self._call_store("compare_set", self._key, token, base64_state_str)
+
+        return self._decode_state(base64_state)
+
+    def _call_store(self, store_op: str, *args, **kwargs) -> Any:
+        try:
+            return getattr(self._store, store_op)(*args, **kwargs)
+        except (ValueError, RuntimeError) as exc:
+            raise RendezvousConnectionError(
+                "The connection to the C10d store has failed. See inner exception for details."
+            ) from exc
+
+    def _decode_state(self, base64_state: bytes) -> Optional[Tuple[bytes, Token]]:
+        if base64_state == self._NULL_SENTINEL.encode():
+            return None
+        try:
+            state = codecs.decode(base64_state, "base64")
+        except binascii.Error as exc:
+            raise RendezvousStateError(
+                "The state object is corrupt. See inner exception for details."
+            ) from exc
+
+        return state, base64_state
+
+
+def _create_tcp_store(params: RendezvousParameters) -> TCPStore:
+    host, port = _parse_rendezvous_endpoint(params.endpoint, default_port=29500)
+
+    cfg_is_host = params.get_as_bool("is_host")
+    # If the user has explicitly specified whether our process should host the
+    # the store, respect it.
+    if cfg_is_host is not None:
+        is_host = cfg_is_host
+    # Otherwise try to determine whether we are the host based on our hostname
+    # and IP address.
+    else:
+        is_host = _matches_machine_hostname(host)
+
+    # The timeout
+    read_timeout = cast(int, params.get_as_int("read_timeout", 60))
+    if read_timeout <= 0:
+        raise ValueError("The read timeout must be a positive integer.")
+
+    # In specific cases we attempt to instantiate the store twice. For details
+    # see the explanation in the except clause below.
+    for _ in range(2):
+        try:
+            store = TCPStore(  # type: ignore[call-arg]
+                host, port, is_master=is_host, timeout=timedelta(seconds=read_timeout)
+            )
+
+            if is_host:
+                log.info(
+                    f"Process {os.getpid()} hosts the TCP store for the C10d rendezvous backend."
+                )
+
+            break
+        except (ValueError, RuntimeError) as exc:
+            # If we heuristically inferred the value of is_host as True and our
+            # first attempt to instantiate the TCP store has failed, try it one
+            # more time with is_host set to False. As an edge case there can be
+            # more than one process that is part of the same rendezvous on this
+            # machine and only one of them will eventually host the store.
+
+            if not is_host or cfg_is_host is not None:
+                raise RendezvousConnectionError(
+                    "The connection to the C10d store has failed. See inner exception for details."
+                ) from exc
+
+            is_host = False
```

Comment:
Wouldn't this be retried by all processes in the world_size with `is_host=False` leading to all of them failing again since no one has `is_host=True`?

Reply:
There are three cases:

- If `is_host` was explicitly provided by the user, we do not re-attempt. After the first attempt, if the store initialization fails, we simply raise an error.
- If we inferred the value of `is_host` to be `False` (our hostname does not match the hostname of the specified rendezvous endpoint), again; we do not re-attempt and raise an error.
- If we inferred the value of `is_host` to be `True` (our hostname matches the hostname of the specified rendezvous endpoint), only then we attempt a second time with `is_master=False` assuming another process on the same machine is also part of the same job and acted faster than us.